### DISCUSSION TITLE
BOS always has the highest probability in logits?

### DISCUSSION THREAD

**Discussion by LoganDark** [ggml-org/llama.cpp#4264] [Q&A] [answered]
I'm trying to build a Rust integration with llama.cpp, but I'm having an issue where everything works fine except that the BOS token (the second value, with index of `1`) always has the highest probability in logits.

![image](https://github.com/ggerganov/llama.cpp/assets/4723091/af6e208d-0d2c-44eb-9832-8930ed9b7073)

This causes it to *always* be picked with greedy sampling, which causes the model to never be able to complete anything because BOS is always empty.

I know I could probably manually exclude this token or use some higher-quality sampling or something, but I've been using [simple.cpp](https://github.com/ggerganov/llama.cpp/blob/1f5cd83275fabb43f2ae92c30033b384a3eb37b4/examples/simple/simple.cpp) as a reference, and I see absolutely no code in there that looks like it could be accounting for this sort of issue.

Is it known why/how this happens and what is the best way to prevent / work around it?

**Comment by LoganDark** [comment, marked as answer]
I found the issue. I was constructing a `llama_context_params` manually, and left zero all of the parameters that I didn't know the purpose of. Well, turns out that some of those parameters have default values that are not zero, so setting them to zero confused llama.cpp.

The solution was to use `llama_context_default_params()` and then only set the individual fields that I actually need to change.

![image](https://github.com/ggerganov/llama.cpp/assets/4723091/81a88a49-d255-4a52-9be9-f9bed7b07236)
